pynn_spinnaker/populations.py

```diff
@@ -250,174 +261,202 @@ def _read_recorded_vars(self, vars_to_read):
+                    _calc_clusters_per_core(max_constraint, c)
+                    for c in current_input_j_constraints)
+                logger.debug("\t\t\t\tMax synapse clusters per current input core:%u",
+                            max_current_input_j)
+                max_clusters = max(max_clusters, max_current_input_j)
+            else:
+                current_input_j_constraints = []
+
+            # Search downwards through possible power-of-two cluster widths
+            logger.debug("\t\t\t\tMax synapse clusters:%u",
+                         max_clusters)
+            max_cluster_power = int(math.log(max_clusters, 2))
+            for s in (2 ** p for p in range(max_cluster_power, -1, -1)):
+                cluster_width = s * max_constraint
+                logger.debug("\t\t\t\tCluster width:%u", cluster_width)
+
+                cluster_synapse_processors = total_syn_processors * s
+                logger.debug("\t\t\t\t\t%u synapse processors",
+                             cluster_synapse_processors)
+
+                cluster_neuron_processors =\
+                    _calc_cores_per_cluster(cluster_width, neuron_j_constraint)
+                logger.debug("\t\t\t\t\t%u neuron processors",
+                             cluster_neuron_processors)
+
+                cluster_current_input_processors = [
+                    _calc_cores_per_cluster(cluster_width, c)
+                    for c in current_input_j_constraints]
+                logger.debug("\t\t\t\t\t%u current input processors",
+                             sum(cluster_current_input_processors))
+
+                # If this configuration can fit on a chip
+                if((cluster_synapse_processors + cluster_neuron_processors +
+                    sum(cluster_current_input_processors)) <= 16):
+                    logger.debug("\t\t\t\t\tCluster fits on chip!")
+
+                    # Calculate final neuron J constraint
+                    self._neuron_j_constraint =\
+                        cluster_width // cluster_neuron_processors
+                    logger.debug("\t\t\t\t\t%u neurons per neuron processor",
+                                 self._neuron_j_constraint)
+
+                    # Calculate final current input J
+                    # constraint for each projection
+                    for p, c in zip(dc_projections, cluster_current_input_processors):
+                        p._current_input_j_constraint = cluster_width // c
+                        logger.debug("\t\t\t\t\t%s - %u neurons per current input processor",
+                                 p.label, p._current_input_j_constraint)
+                    return
+
+            # Divide the constraint on any synapse processors
+            # which currently have maximum constraint by 2
+            new_max_constraint = max_constraint // 2
+            self._synapse_j_constraints =\
+                {s_type : new_max_constraint if c == max_constraint else c
```

Comment:
Add comment clarifying that ``c <= max_constraint`` for all c and rename c to constraint